services/data/src/react/hooks/useDataMutation.ts

```diff
@@ -1,32 +1,79 @@
-import { useCallback } from 'react'
-import { QueryOptions, Mutation } from '../../engine'
+import { useState, useCallback } from 'react'
+import { useMutation } from 'react-query'
+import { QueryVariables, QueryOptions, Mutation } from '../../engine'
 import { MutationRenderInput } from '../../types'
 import { useDataEngine } from './useDataEngine'
-import { useQueryExecutor } from './useQueryExecutor'
 import { useStaticInput } from './useStaticInput'
 
-const empty = {}
 export const useDataMutation = (
     mutation: Mutation,
-    { onComplete, onError, variables = empty, lazy = true }: QueryOptions = {}
+    {
+        onComplete: userOnComplete,
+        onError: userOnError,
+        variables: initialVariables = {},
+        lazy = true,
+    }: QueryOptions = {}
 ): MutationRenderInput => {
     const engine = useDataEngine()
+    const [variables, setVariables] = useState(initialVariables)
+    const [shouldMutateNow, setShouldMutateNow] = useState(!lazy)
     const [theMutation] = useStaticInput<Mutation>(mutation, {
         warn: true,
         name: 'mutation',
     })
-    const execute = useCallback(
-        options => engine.mutate(theMutation, options),
-        [engine, theMutation]
-    )
-    const { refetch: mutate, called, loading, error, data } = useQueryExecutor({
-        execute,
-        variables,
-        singular: false,
-        immediate: !lazy,
-        onComplete,
+
+    // Ensure we call the supplied onComplete only with the data
+    const onSuccess = (data: any) => {
+        userOnComplete && userOnComplete(data)
+    }
+    // Ensure we call the supplied onError only with the error
+    const onError = (error: any) => {
+        userOnError && userOnError(error)
+    }
+
+    const mutationFn = (newVariables: QueryVariables): Promise<any> => {
+        const mergedVariables = { ...variables, ...newVariables }
+
+        setVariables(mergedVariables)
+        return engine.mutate(theMutation, { variables: mergedVariables })
+    }
+
+    const result = useMutation(mutationFn, {
+        onSuccess,
         onError,
     })
 
-    return [mutate, { engine, called, loading, error, data }]
+    if (shouldMutateNow) {
+        setShouldMutateNow(false)
+
+        // Not passing variables since we only have initialVariables at this point
+        result.mutate({})
+    }
```

Comment:
Why using state + condition in render code?

```js
useEffect(() => {
  if (lazy) {
    result.mutate({})
  }
}, [])
```

Reply:
Ah of course yeah, that's a lot better. No idea why I didn't think of that. I'll update it!

Reply:
Fixed in c47c2e43b0277d7c51d13d6eaa2ee76e9be6ca7a